### PR TITLE
Add ERC20 support

### PR DESCRIPTION
**Pull request by lalexgap**
Both the L1 and L2 contracts now support the use of ERC20 tokens.

Both the `Ticket` and `EscrowEntry` structs and types now include a `token` which is either:
- An address of a ERC20 token
- `address(0)` indicating that ETH should be transferred.

This `token` field is then used throughout the contracts to determine if eth or a token need to be transferred.

To enable the use of ERC20 tokens in tests the `USE_ERC20` env var must be to `true`.  I've done this because the `hardhat-gas-reporter` seems to print out a gas summary for all tests. By having a flag it makes comparisons easier between ERC20 token gas costs and native gas costs.

### CODE REVIEW

File: contracts/L1.sol [hunk 1]
```diff
@@ -51,19 +53,26 @@ contract L1Contract is SignatureChecker {
...
 
-    /// Used by Bob to deposit funds to fund various tickets.
-    function deposit() public payable {
-        // EFFECTS
-        balances[msg.sender] += msg.value;
+    function depositEth() public payable {
+        balances[msg.sender][address(0)] += msg.value;
+    }
+
+    function depositToken(address token, uint256 amount) public {
```
Comment (lalexgap) [inline contracts/L1.sol:70]: We could just overload the deposit function but its annoying to deal with using `ethers.js`

**Comment by andrewgordstewart** [comment]
Screen Shot 2021-12-01 at 2 03 42 PM


How hard is it to display the average cost of a swap in the CI output? It'd be nice to
- quickly see the difference in cost between ERC20 and eth
- allow the developer and reviewer to observe the effect of the change on the average cost of a swap

**Comment by lalexgap** [comment]
> How hard is it to display the average cost of a swap in the CI output? It'd be nice to
> 
> * quickly see the difference in cost between ERC20 and eth
> * allow the developer and reviewer to observe the effect of the change on the average cost of a swap

I think we'd have to do our own gas accounting, but I'm not sure how much effort that would be. @geoknee may have some good insight into this?

**Comment by geoknee** [comment]
I estimate probably 0.5 to 1 dev day to get our own gas accounting set up, following what we did [here](https://github.com/statechannels/statechannels/tree/master/packages/nitro-protocol/gas-benchmarks). I think it would be well worth the effort.

We could reuse quite a bit of the code that [we already wrote](https://github.com/statechannels/statechannels/blob/55ab6fd28ab1bade93a7d9648b25a8b63bd9dd55/packages/nitro-protocol/gas-benchmarks/vanillaSetup.ts#L69).

**Review by kerzhner — APPROVED** [review]

**Pull request merged**